### DISCUSSION TITLE
Lab 02.04 - Task 3: Run the card - Solucionado

### DISCUSSION THREAD

**Discussion by natandreli** [rramosp/ai4eng.v1#225] [02 - PYTHON]
Buenas tardes:

Solucioné la task 3 del Lab 02.04, cuando lo corro manualmente todo sale perfecto:

![Captura de Pantalla 2023-02-19 a la(s) 12 22 29 p m](https://user-images.githubusercontent.com/110438161/219964101-2bbc180a-05fe-4887-98bf-79f378f75b1b.png)

Sin embargo, al evaluarlo en el grader me sale el siguiente error:

![Captura de Pantalla 2023-02-19 a la(s) 12 23 27 p m](https://user-images.githubusercontent.com/110438161/219964151-d7a5fa21-ad3e-4ab6-b812-ead05287f662.png)

Ya se utiliza un assert en la función make_heaps(), igualmente lo agregué a la función run() por si este era el problema.

Les dejo el bloque de código de la función run():

```
def run(c, n, n_heaps=3):
  assert n_heaps%2==1, "must have an odd number of heaps"
  assert len(c)%n_heaps==0, "the length of the deck must be a multiple of the number of heaps"
  for i in range(0, 3):
    h = make_heaps(c, n_heaps=3)
    c = collect_heaps(h, n)
  
  r = c.index(n)
  return r
```

Gracias, espero me puedan ayudar.

**Comment by imEag** [comment]
Hola Natalia! EL error se está generando en la función run a la hora de llamar a la función make_heaps. Observa que a la hora de llamar una función podemos pasarle argumentos con la notación `nombre_del_argumento=valor` en este caso estas asignando `n_heaps=3` lo que genera que los heaps siempres se generen con este parámetro en 3 sin importar el valor que haya recibido la función run en ese argumento. Lo que deberías estar haciendo es `n_heaps=n_heaps` para tener el cuenta el valor que recibe la función run

**Reply by natandreli** [reply]
¡Hola! Ya lo corregí y me sigue apareciendo el mismo error. 

Sin embargo, pienso que el igualar el argumento n_heaps a 3 no tiene que representar algún error, ya que sería un valor por defecto que se le daría al parámetro, es decir, n_heaps toma el valor de 3 sólo en el caso en el que no se le pase otro valor al momento de llamar la función. 

Te dejo un ejemplo que ilustra lo que quiero decir, encontrado en línea:

![Captura de Pantalla 2023-02-19 a la(s) 1 45 24 p m](https://user-images.githubusercontent.com/110438161/219968556-327085ea-c780-4082-95aa-0b4be405465e.png)

Gracias igualmente, seguiré pendiente.

**Reply by imEag** [reply]
igualar el parámetro de una función para dar un valor por defecto solo funciona a la hora de definir una función, por ejemplo: 
`
def my_function(optional_param=12):
    ....
`
En este caso lo que se hace es que si optional_param no recibe un valor se asigna automáticamente el 12.

Cuando se llama la función, si ponemos el argumento y lo igualamos a algo, lo que estamos haciendo es dándole un valor que va a tomar. Ejemplo: 
`my_fuction(optional_param=30)`
En este caso lo que se hace es darle el valor de 30 a optional_param en este llamado a la función

**Reply by natandreli** [reply]
¡Vale! Ya vi a qué te referías, estoy pasando el 3 al momento de llamar la función dentro del ciclo.

¡Muchas gracias!

**Reply by imEag** [reply]
muestrame por favor el código implementando lo que te sugerí y el nuevo mensaje de error, gracias

**Reply by imEag** [reply]
exacto, a eso me refería inicialmente

**Reply by natandreli** [reply]
@imEag muchísimas gracias ☺️

**Reply by imEag** [reply]
Con mucho gusto

**Comment by Felipe-RA** [comment]
Saludos @natandreli 

Me gustaría ver tu implementación de `make_heaps()` y de `collect__heaps()` para poderte dar un consejo más preciso sobre una posible solución a tu implementación.

Gracias!

**Reply by natandreli** [reply]
Claro que sí, mira:

```
def make_heaps(c, n_heaps=3):
    assert n_heaps%2==1, "must have an odd number of heaps"
    assert len(c)%n_heaps==0, "the length of the deck must be a multiple of the number of heaps"
    
    n_cards = int(len(c)/n_heaps)
    h = []

    for j in range(0, n_heaps):
      l_cards = []
      for i in range(0, n_cards):
        l_cards.append(c[i*n_heaps + j])
      h.append(l_cards)

    return h
```

```
def collect_heaps(h, n):

  for i in range(0, len(h)):
    if n in h[i]:
      p_card = i

  if p_card >= len(h)//2:
    h.insert(len(h)//2, h[p_card])
    h.pop(p_card + 1)
  else:
    h.insert(len(h)//2 + 1, h[p_card])
    h.pop(p_card)

  r = []

  for i in range(0, len(h)):
    r.extend(h[i])

  return r            
```

¡Muchas gracias!

**Reply by natandreli** [reply]
¡Ya encontré mi error! Dentro del ciclo estoy pasando un argumento igualado a 3, por eso no me funciona.

Muchas gracias igualmente.
